quickjs.c

```diff
@@ -47251,145 +47251,413 @@ static JSValue js_Date_UTC(JSContext *ctx, JSValue this_val,
+            return FALSE;
+        n = p - n;
+        if (strict && n != 2 && n != 4)
+            return FALSE;
+        while (n > 4) {
+            n -= 2;
+            hh /= 100;
+        }
+        if (n > 2) {
+            mm = hh % 100;
+            hh = hh / 100;
+        } else {
+            mm = 0;
+            if (string_skip_char(sp, &p, ':')  /* optional separator */
+            &&  !string_get_digits(sp, &p, &mm, 2, 2))
+                return FALSE;
+        }
+        if (hh > 23 || mm > 59)
+            return FALSE;
+        tz = hh * 60 + mm;
+        if (sgn != '+')
+            tz = -tz;
+    } else
+    if (sgn != 'Z') {
+        return FALSE;
     }
-    *pval = v;
     *pp = p;
-    return 0;
+    *tzp = tz;
+    return TRUE;
 }
 
-static int string_get_milliseconds(JSString *sp, int *pp, int64_t *pval) {
-    /* parse milliseconds as a fractional part, round to nearest */
-    /* XXX: the spec does not indicate which rounding should be used */
-    int mul = 1000, ms = 0, p = *pp, c, p_start;
-    if (p >= sp->len)
-        return -1;
-    p_start = p;
-    while (p < sp->len) {
-        c = string_get(sp, p);
-        if (!(c >= '0' && c <= '9')) {
-            if (p == p_start)
-                return -1;
-            else
-                break;
-        }
-        if (mul == 1 && c >= '5')
-            ms += 1;
-        ms += (c - '0') * (mul /= 10);
+static uint8_t upper_ascii(uint8_t c) {
+    return c >= 'a' && c <= 'z' ? c - 'a' + 'Z' : c;
+}
+
+static BOOL string_match(const uint8_t *sp, int *pp, const char *s) {
+    int p = *pp;
+    while (*s != '\0') {
+        if (upper_ascii(sp[p]) != upper_ascii(*s++))
+            return FALSE;
         p++;
     }
-    *pval = ms;
     *pp = p;
-    return 0;
+    return TRUE;
 }
 
-
-static int find_abbrev(JSString *sp, int p, const char *list, int count) {
+static int find_abbrev(const uint8_t *sp, int p, const char *list, int count) {
     int n, i;
 
-    if (p + 3 <= sp->len) {
-        for (n = 0; n < count; n++) {
-            for (i = 0; i < 3; i++) {
-                if (string_get(sp, p + i) != month_names[n * 3 + i])
-                    goto next;
-            }
-            return n;
-        next:;
+    for (n = 0; n < count; n++) {
+        for (i = 0;; i++) {
+            if (upper_ascii(sp[p + i]) != upper_ascii(list[n * 3 + i]))
+                break;
+            if (i == 2)
+                return n;
         }
     }
     return -1;
 }
 
-static int string_get_month(JSString *sp, int *pp, int64_t *pval) {
+static BOOL string_get_month(const uint8_t *sp, int *pp, int *pval) {
     int n;
 
-    string_skip_spaces(sp, pp);
     n = find_abbrev(sp, *pp, month_names, 12);
     if (n < 0)
-        return -1;
+        return FALSE;
 
-    *pval = n;
+    *pval = n + 1;
     *pp += 3;
-    return 0;
+    return TRUE;
+}
+
+/* parse toISOString format */
+static BOOL js_date_parse_isostring(const uint8_t *sp, int fields[9], BOOL *is_local) {
```

Comment:
```suggestion
static BOOL js_date_parse_isostring(const uint8_t *sp, int (*fields)[9], BOOL *is_local) {
```
That way callers _must_ pass in an exactly nine-element array (and you have to dereference them here as `(*fields)[i]`)

Another alternative is C99-style `int fields[static 9]` syntax. Then callers must pass in an at least nine-element array.

Reply:
Good suggestion.  I don't like the C99 syntax too much.  Does MSAN or ASAN catch out of bound accesses for arrays passed as pointers (without a length specified by either syntax)?

Reply:
Most of the time. Not 100% of the time though.

Reply:
I added a macro to encapsulate the C99 syntax and used that. It is less surprising to casual readers than the array pointer trick, and also allows for larger arrays if necessary, as is the case for `get_date_fields()` / `set_date_fields()` combinations.

Reply:
I had to disable the C99 syntax for MSVC that does not support it even though it does define `__STDC_VERSION__`.

Reply:
The `int (*fields)[9]` syntax works in msvc; it's a c89-ism that we use quite extensively in libuv.

Reply:
> The `int (*fields)[9]` syntax works in msvc; it's a c89-ism that we use quite extensively in libuv.

Of course, but this syntax would force the length to be `9` whereas `get_date_fields` needs at least `9` and  `set_date_fields` at least `7`.  It is not a problem if MSVC cannot verify the minimum length as other compilers already do for 99,9% of the code.